corehq/apps/auditcare/tests/test_models.py

```diff
@@ -78,6 +84,20 @@ def test_audit_view_should_not_save(self):
         self.assertIsNone(event.id)
 
 
+def test_get_domain():
+    def test(cfg):
+        request = make_request(cfg.path)
+        if "request_domain" in cfg:
+            request.domain = cfg.request_domain
+        eq(mod.get_domain(request), cfg.expect)
+
+    cfg = Config(expect="block")
+    yield test, cfg(path="/path", expect=None)
+    yield test, cfg(path="/a/block/path")
+    yield test, cfg(path="/path", request_domain="block")
+    yield test, cfg(path="/a/block/path", request_domain="xx")
```

Comment:
The config argument is represented as part of the test name, so in the case of a test failure it is easy to see the expected value as well as the inputs.
```
FAIL: corehq.apps.auditcare.tests.test_models:test_get_domain ((expect='block', path='/a/block/path'),)
```

To me, it's a succinct way of testing multiple combinations of arguments. It was written in the spirit of the best practice outlined in _Software Engineering at Google_: "require the test author to specify only values they care about, and setting reasonable defaults for all other values."

Reply:
Sorry, indulge me for a wall-of-quotes from the same chapter.


>When a test fails, an engineer’s first job is to identify which of these cases the failure falls into and then to diagnose the actual problem. The speed at which the engineer can do so depends on the test’s clarity. A clear test is one whose purpose for existing and reason for failing is immediately
clear to the engineer diagnosing a failure. Tests fail to achieve clarity when their reasons for failure aren’t obvious or when it’s difficult to figure out why they were originally written. Clear tests also bring other benefits, such as documenting the system under test and more easily serving as a basis for new tests.

>For a test suite to scale and be useful over time, it’s important that each individual test in that suite be as clear as possible.

If any of these tests failed, the intent of each test is not immediately apparent.

>A test is complete when its body contains all of the information a reader needs in order to understand how it arrives at its result. A test is concise when it contains no other distracting or irrelevant information... a test's body should contain all the information needed to understand it without containing any irrelevant or distracting information

This is the basis for the quote you cited -- avoiding extraneous information. This helper does the opposite -- it obscures the expectation. If we needed to provide a port to make our request, that would be a great thing to hide in a helper method, because it's irrelevant to what we're testing; conversely, the expectation is critical to what we're testing.

>Remember that every behavior has three parts: a "given" component that defines how the system is set up, a "when" component that defines the action to be taken on the system, and a "then" component that validates the result. Tests are clearest when this structure is explicit.

>The test name is very important: it will often be the first or only token visible in failure reports, so it's your best opportunity to communicate the problem when the test breaks. It's also the most straightforward way to express the intent of the test.

>A good name describes both the actions that are being taken on a system and the expected outcome.

>Names like this are much more verbose than we’d normally want to write for methods in production code, but the use case is different: we never need to write code that calls these, and their names frequently need to be read by humans in reports. Hence, the extra verbosity is warranted.

Parameterized tests like these lose all of this information, and thus become less clear. Here's my understanding of what each is testing (this wasn't immediately apparent to me):
- When a path contains no domain, None should be returned: `path_lacking_domain_returns_None`
- When a path contains a domain, that domain should be returned: `path_including_domain_returns_domain`
- When a request contains both a path and an explicit domain, the explicit domain is used if the path does not contain one: `path_lacking_domain_returns_domain_param_when_domain_param_exists`
- When a request contains both a path with a domain and an explicit domain, the path's domain is used: `path_including_domain_is_used_over_domain_param`

>in test code, stick to straight-line code over clever logic, and consider tolerating some duplication when it makes the test more descriptive and meaningful.

>Clear tests are trivially correct upon inspection.

>These tests have more duplication, and the test bodies are a bit longer, but the extra verbosity is worth it. Each individual test is far more meaningful and can be understood entirely without leaving the test body. A reader of these tests can feel confident that the tests do what they claim to do and aren’t hiding any bugs.

Verbosity isn't a bad thing with tests -- logic in code generally is. These tests, ideally, will be written once and won't change.